platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfiguration.java

```diff
@@ -1347,35 +1391,47 @@ public ConfigurationIdentity getConfigurationIdentity() {
         return new DefaultConfigurationIdentity(buildPath, projectPath, name);
     }
 
-    private boolean isProperUsage(boolean allowDeprecated, ProperMethodUsage... properUsages) {
-        for (ProperMethodUsage properUsage : properUsages) {
-            if (properUsage.isProperUsage(this, allowDeprecated)) {
-                return true;
-            }
-        }
-        return false;
+    private boolean isProperUsage(ProperMethodUsage... properUsages) {
+        ConfigurationInternal conf = this;
+        return Arrays.stream(properUsages).anyMatch(pu -> pu.isAllowed(conf));
+    }
+
+    private boolean isDeprecatedUsage(ProperMethodUsage... properUsages) {
+        ConfigurationInternal conf = this;
+        return Arrays.stream(properUsages).anyMatch(pu -> pu.isDeprecated(conf));
     }
 
-    private void warnOnInvalidInternalAPIUsage(String methodName, ProperMethodUsage... properUsages) {
-        warnOnDeprecatedUsage(methodName, true, properUsages);
+    private void warnOrFailOnInvalidInternalAPIUsage(String methodName, ProperMethodUsage... properUsages) {
+        warnOrFailOnInvalidUsage(methodName, true, properUsages);
     }
 
-    private void warnOnDeprecatedUsage(String methodName, ProperMethodUsage... properUsages) {
-        warnOnDeprecatedUsage(methodName, false, properUsages);
+    private void warnOrFailOnInvalidUsage(String methodName, ProperMethodUsage... properUsages) {
+        warnOrFailOnInvalidUsage(methodName, false, properUsages);
     }
 
-    private void warnOnDeprecatedUsage(String methodName, boolean allowDeprecated, ProperMethodUsage... properUsages) {
-        if (!isProperUsage(allowDeprecated, properUsages)) {
-            String msgTemplate = "Calling configuration method '%s' is deprecated for configuration '%s', which has permitted usage(s):\n" +
-                "%s\n" +
-                "This method is only meant to be called on configurations which allow the %susage(s): '%s'.";
+    private void warnOrFailOnInvalidUsage(String methodName, boolean allowDeprecated, ProperMethodUsage... properUsages) {
+        if (!isProperUsage(properUsages)) {
             String currentUsageDesc = UsageDescriber.describeCurrentUsage(this);
             String properUsageDesc = ProperMethodUsage.summarizeProperUsage(properUsages);
 
-            DeprecationLogger.deprecateBehaviour(String.format(msgTemplate, methodName, getName(), currentUsageDesc, allowDeprecated ? "" : "(non-deprecated) ", properUsageDesc))
-                .willBeRemovedInGradle9()
-                .withUpgradeGuideSection(8, "deprecated_configuration_usage")
-                .nagUser();
+            if (isDeprecatedUsage(properUsages)) {
```

Comment:
I would expect `isDeprecatedUsage` to be outside of the `!isProperUsage` if statement. Currently, say for some resolvable but deprecated for resolution configuration, `isProperUsage` should return true since `configuration.isCanBeResolved()` is true, meaning `!isProperUsage` is false, and we would never check if the configuration is deprecated. 

To me it should be like:
```
if (!isProperUsage) {
    // throw exception
} else if (isDeprecated) {
    // emit warning
}
```